form/root_storage/root_ttree_container.cpp

```diff
@@ -0,0 +1,149 @@
+                                                                     {"double", "/D"},
+                                                                     {"short int", "/S"},
+                                                                     {"unsigned short", "/s"},
+                                                                     {"long int", "/L"},
+                                                                     {"unsigned long int", "/l"},
+                                                                     {"bool", "/O"}};
+}
+
+ROOT_TTree_ContainerImp::ROOT_TTree_ContainerImp(const std::string& name) :
+  m_name(name),
+  m_treeName(""),
+  m_branchName(""),
+  m_file(nullptr),
+  m_tree(nullptr),
+  m_ownsTree(false),
+  m_branch(nullptr)
+{
+  auto del_pos = m_name.find("/");
+  if (del_pos != std::string::npos) {
+    m_treeName = m_name.substr(0, del_pos);
+    m_branchName = m_name.substr(del_pos + 1);
+  } else {
+    m_treeName = m_name;
+    m_branchName = m_name;
+  }
+}
+
+ROOT_TTree_ContainerImp::~ROOT_TTree_ContainerImp()
+{
+  if (m_ownsTree && m_tree != nullptr) {
+    m_tree->Write();
+    delete m_tree;
+  }
+}
+
+const std::string& ROOT_TTree_ContainerImp::name() { return m_name; }
+
+void ROOT_TTree_ContainerImp::setFile(std::shared_ptr<IStorage_File> file)
+{
+  m_file = dynamic_cast<ROOT_TFileImp*>(file.get())->getTFile();
+  return;
+}
+
+void ROOT_TTree_ContainerImp::fill(const void* data, const std::string& type)
+{
+  if (m_file == nullptr) {
+    throw std::string("ROOT_TTree_ContainerImp::fill no file attached");
+  }
+  if (m_tree == nullptr) {
+    m_tree = m_file->Get<TTree>(m_treeName.c_str());
+  }
+  if (m_tree == nullptr) {
+    m_tree = new TTree(m_treeName.c_str(), m_treeName.c_str());
+    m_tree->SetDirectory(m_file.get());
+    m_ownsTree = true;
+  }
+  if (m_tree == nullptr) {
+    throw std::string("ROOT_TTree_ContainerImp::fill no tree created");
+  }
+  if (m_branch == nullptr) {
+    m_branch = m_tree->GetBranch(m_branchName.c_str());
+  }
+
+  auto dictInfo = TDictionary::GetDictionary(type.c_str());
+  if (m_branch == nullptr) {
+    if (!dictInfo) {
+      throw std::string("Unsupported type passed to ROOT_TTree_ContainerImp::fill: " + type);
+    }
+    if (dictInfo->Property() & EProperty::kIsFundamental) {
+      m_branch = m_tree->Branch(m_branchName.c_str(),
+                                &data,
+                                (m_branchName + typeNameToLeafList[dictInfo->GetName()]).c_str(),
+                                4096);
+    } else {
+      m_branch = m_tree->Branch(m_branchName.c_str(), dictInfo->GetName(), &data);
+    }
+  }
+  if (m_tree == nullptr) {
+    throw std::string("ROOT_TTree_ContainerImp::fill no branch created");
+  }
+  if (dictInfo->Property() & EProperty::kIsFundamental) {
+    m_branch->SetAddress(const_cast<void*>(data)); //FIXME: const_cast?
+  } else {
+    m_branch->SetAddress(&data);
+  }
+  m_branch->Fill();
```

Comment:
The pattern in this function implies the filling of the branches might be delayed ... unless there is some mechanism in place to ensure that all branches are always started to be filled at the same time, we might need to be back filling this branch to catch up with the other branches (i.e. having the same number of entries) (To back fill call `TBranch::BackFill()` after setting the data to have the 'default' value.

Reply:
Kyle confirmed all algorithm invocations write same data products -> same branches, no need to BackFill().